### PR TITLE
Impl UUID macro

### PR DESCRIPTION
**Pull request by QnnOkabayashi**
**I'm submitting a feature**

# Description
Implementing the `uuid!` macro for parsing string literals at compile time.

# Motivation
Very convenient for config files

# Tests
I duplicated the valid str parsing tests for `uuid!()`, and they all compiled. For testing for failure, I checked that invalid inputs don't work:

1. non literal: `uuid!(struct)` fails properly.
2. non string literal: `uuid!(44)` fails properly.
3. invalid string: `uuid!("hello world")` fails properly

I didn't bother to test more extensively then that, do we care?

Also, can someone verify that tokenizing to `::uuid::Uuid::from_bytes([#tokens])` is hygienic?

# Related Issue(s)
https://github.com/uuid-rs/uuid/issues/422

### CODE REVIEW

File: macros/src/lib.rs [hunk 1]
```diff
@@ -1,17 +1,106 @@
...
+use std::fmt;
+use syn::spanned::Spanned;
+
+#[cfg(any(feature = "std", test))]
+#[macro_use]
+extern crate std;
+
+#[cfg(all(not(feature = "std"), not(test)))]
+#[macro_use]
+extern crate core as std;
```
Comment (QnnOkabayashi) [inline macros/src/lib.rs:13]: Required to stop `shared/error.rs` and `shared/parser.rs` from complaining. I just copy-pasted it from `uuid`.

**Comment by KodrAus** [comment]
Thanks for working on this @QnnOkabayashi!

I think we might need to wrap this up in a `macro_rules` macro in `uuid` that can pass the fully-qualified path to the `Uuid` type using the `$crate` metavariable. We could then move the docs to this public macro and make the proc-macro a private implementation detail.

It would also be great to test this using `trybuild` so we can look at diagnostics for invalid inputs, and test that renaming `uuid` or defining a root module called `uuid` doesn’t cause the macro to break.

Is this something you’d be interested in exploring further? If not I’m happy to merge and fill those bits in.

Thanks again! 🙇

**Comment by QnnOkabayashi** [comment]
> I think we might need to wrap this up in a `macro_rules` macro in `uuid` that can pass the fully-qualified path to the `Uuid` type using the `$crate` metavariable. We could then move the docs to this public macro and make the proc-macro a private implementation detail.

Done 👍 
> 
> It would also be great to test this using `trybuild` so we can look at diagnostics for invalid inputs, and test that renaming `uuid` or defining a root module called `uuid` doesn’t cause the macro to break.

Done 👍 

Also since the macro is integrated with the crate now, I was able to use the internal diagnostic information in the `ErrorKind` type to improve the diagnostics. For example, if there's an invalid character, it only points to that character
```
error: invalid character: expected an optional prefix of `urn:uuid:` followed by 0123456789abcdefABCDEF-, found G at 20
  --> benches/macros/invalid_parse.rs:25:44
   |
25 | const _: Uuid = uuid!("F9168C5E-CEB2-4faa-BGBF-329BF39FA1E4");
   |                                            ^
```
If there's an invalid group length, it points to that group
```
error: invalid group length: expected 12, found 8 in group 4
  --> benches/macros/invalid_parse.rs:26:48
   |
26 | const _: Uuid = uuid!("01020304-1112-2122-3132-41424344");
   |                                                ^^^^^^^^
```

**Review by KodrAus — APPROVED** [review]
This looks great, thanks @QnnOkabayashi!

I should've merged this one ahead of my changes to our benches. I just renamed some of the files. We probably don't really need benches for the `uuid!` macro, since it should be "free" 🙂

**Comment by KodrAus** [comment]
Besides the merge conflict this should be ready to go now.

**Comment by QnnOkabayashi** [comment]
> Besides the merge conflict this should be ready to go now.

Should be good to go now!

**Pull request merged**